src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs

```diff
@@ -0,0 +1,270 @@
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(_database.DatabaseShutdown);
+			_heartbeatMessage = _context.ReadObject(new DynamicJsonValue
+			{
+				[Constants.MessageType] = "Heartbeat"
+			}, "heartbeat msg");
+		}
+
+        public void Start()
+        {
+            _sendingThread = new Thread(ReplicateDocuments)
+            {
+                Name = "Replication from " + _database.Name + " to remote " + _destination.Database + " at " + _destination.Url,
+                IsBackground = true
+            };
+            _sendingThread.Start();
+        }
+
+
+	    private void ReplicateDocuments()
+        {
+	        try
+	        {
+		        var connectionInfo = GetTcpInfo();
+		        using (var tcpClient = new TcpClient())
+		        {
+			        ConnectSocket(connectionInfo, tcpClient);
+			        using (var stream = tcpClient.GetStream())
+			        using (var writer = new BlittableJsonTextWriter(_context, stream))
+			        {
+				        _context.Write(writer, new DynamicJsonValue
+				        {
+					        ["DatabaseName"] = _database.Name,
+					        ["Operation"] = TcpConnectionHeaderMessage.OperationTypes.Replication.ToString(),
+					        ["DatabaseId"] = _database.DbId.ToString(),
+				        });
+
+				        _context.Write(writer, new DynamicJsonValue
+				        {
+					        [Constants.MessageType] = Constants.Replication.MessageTypes.GetLastEtag
+				        });
+				        writer.Flush();
+
+				        using (var lastEtagMessage = _context.ReadForMemory(stream, "Last etag from server"))
+				        {
+					        var replicationEtagReply = JsonDeserialization.ReplicationEtagReply(lastEtagMessage);
+					        _lastSentEtag = replicationEtagReply.LastSentEtag;
+				        }
+
+				        while (_cts.IsCancellationRequested == false)
+				        {
+					        if (!ExecuteReplicationOnce(writer, stream))
+						        using (_context.OpenReadTransaction())
+							        if (DocumentsStorage.ReadLastEtag(_context.Transaction.InnerTransaction) < _lastSentEtag)
+								        continue;
+
+					        //if this returns false, this means either timeout or canceled token is activated                    
+							while (_waitForChanges.Wait(_minimalHeartbeatInterval, _cts.Token) == false)
+						        SendHeartbeat(stream);
+				        }
+			        }
+		        }
+	        }
+	        catch (OperationCanceledException)
+	        {
+				if (_log.IsInfoEnabled)
+					_log.Info($"Operation canceled on replication thread ({FromToString}). Stopped the thread.");
+	        }
+            catch (Exception e)
+            { 
+				if(_log.IsInfoEnabled)
+					_log.Info($"Unexpected exception occured on replication thread ({FromToString}). Stopped the thread.", e);
+                Failed?.Invoke(this, e);
+            }
+        }
+
+	    private string FromToString => $"from {_database.ResourceName} to {_destination.Database} at {_destination.Url}";
```

Comment:
Need to use this consistently, for example, SendHeartbeat isn't using it, nor it the thread name